Python/compile.c

```diff
@@ -1622,6 +1633,120 @@ compiler_unwind_fblock_stack(struct compiler *c, location *ploc,
     return SUCCESS;
 }
 
+static int
+compiler_setup_annotations_scope(struct compiler *c, location loc,
+                                 void *key, PyObject *name)
+{
+    if (compiler_enter_scope(c, name, COMPILER_SCOPE_ANNOTATIONS,
+                             key, loc.lineno) == -1) {
+        return ERROR;
+    }
+    c->u->u_metadata.u_posonlyargcount = 1;
+    // if .format != 1: raise NotImplementedError
+    _Py_DECLARE_STR(format, ".format");
+    ADDOP_I(c, loc, LOAD_FAST, 0);
+    ADDOP_LOAD_CONST(c, loc, _PyLong_GetOne());
+    ADDOP_I(c, loc, COMPARE_OP, (Py_NE << 5) | compare_masks[Py_NE]);
+    NEW_JUMP_TARGET_LABEL(c, body);
+    ADDOP_JUMP(c, loc, POP_JUMP_IF_FALSE, body);
+    ADDOP_I(c, loc, LOAD_COMMON_CONSTANT, CONSTANT_NOTIMPLEMENTEDERROR);
+    ADDOP_I(c, loc, RAISE_VARARGS, 1);
+    USE_LABEL(c, body);
+    return 0;
+}
+
+static int
+compiler_leave_annotations_scope(struct compiler *c, location loc,
+                                 Py_ssize_t annotations_len)
+{
+    ADDOP_I(c, loc, BUILD_MAP, annotations_len);
+    ADDOP_IN_SCOPE(c, loc, RETURN_VALUE);
+    PyCodeObject *co = optimize_and_assemble(c, 1);
+    compiler_exit_scope(c);
+    if (co == NULL) {
+        return ERROR;
+    }
+    if (compiler_make_closure(c, loc, co, 0) < 0) {
+        Py_DECREF(co);
+        return ERROR;
+    }
+    Py_DECREF(co);
+    return 0;
+}
+
+static int
+compiler_collect_annotations(struct compiler *c, asdl_stmt_seq *stmts,
+                             int *annotations_len)
+{
```

Comment:
Do we really need to add a second pass to the compiler for this?

Reply:
The alternative would I think be to set up some system where we are emitting instructions into two different code objects at the same time, which would likely be more complicated. Note that this second pass happens only on module and class bodies and does not recurse into AST nodes that cannot contain AnnAssign statements. I believe I also made it so the pass runs only if the symtable has detected that there are any annotations in this block.

Reply:
I think there is another alternative - to collect the annotations into a list on the compiler unit during the normal pass, and then use this list to create the annotations code object. Would that not work?

I'm not concerned about performance, but about maintenance complexity. A second pass is just asking to get out of sync when we change the AST node set (that it's only considering some of the node types makes this more likely).

Reply:
That's a reasonable idea. We in fact already have a pass like this in the `find_ann()` functions, and we've had some past bugs where annotations were dropped if they appeared inside some blocks because they were missed in `find_ann()`. I realized it's easy to get rid of this pass by recording some data in symtable.c; I'll push a commit for that.

However, the case I'm adding in this PR is a bit harder. AST nodes aren't PyObjects, so I can't put them in a list. I suppose I could convert AnnAssign pointers into Python ints and then stash them in a list.